### ISSUE TITLE
Session data is not being stored (Inconsistent)

### ISSUE THREAD

**Issue by deveshMantra** [expressjs/session#710]
Trying to save extra data in session and redirecting to the other page after coming back the extra data is not stored rest of the thing in session is available this is happening inconsistently.

server.js file
```
```app.use(cookieParser())
app.use(helmet())
app.use(session({
  secret: 'some secret key',
  resave: false,
  saveUninitialized: false,
  store: memoryStore
}))```
```

once the user logs in the system we are storing the following data in session.
```
```{ cookie: { originalMaxAge: null, expires: null, httpOnly: true, path: '/' },
  auth_redirect_uri: 'http://localhost:3000/resources?auth_callback=1',
  'keycloak-token': '{"access_token":"someToken","expires_in":86400,"refresh_expires_in":43200,"refresh_token":"someToken","not-before-policy":0,"session_state":"89e0e02e-1414-4043-ba1b-a61f23e5dfa4"}',
  userId: '2752-4e2e-bb6d-d6693c3190d5',
} ```

```

User tries another api we are storing more data into the existing session.

` mergeAccountInfo: 
   { initiatorAccountDetails: 
      { userId: '9b7de845-2752-4e2e-bb6d-d6693c3190d5',
        sessionToken: 'some token',
        redirectUri: '/resources',
        test: 'mockValue' } }`

To save the session below is code I am trying to use 

```
req.session.newDataToSave = {'some data': ' some more data'}
   req.session.save(function (result) {
    const url = redirect to different url;
    res.redirect(url)
    });
```

after i come back to the redirect url i see the new data that we just saved is not present while the other data is present that was saved after user logged in to system.

Any help on same is highly appreciable @beeman @voxpelli @mikeal @juliangruber @tj

**Comment by HarshithaKP** [comment]
@deveshMantra I have written a simple example to demonstrate session persistence with redirection here : https://gist.github.com/HarshithaKP/3d9ad81138245aa7527bf385d37daba7 
It is working fine for me, let me know if it helps you.

**Comment by gireeshpunathil** [comment]
@deveshMantra - was your issue resolved with the example?

**Comment by deveshMantra** [comment]
@gireeshpunathil 
The issue is resolved in my code I did a hack for the same as it was production issue but I think we need to find another solution in the package itself to avoid such conditions.
There was a race condition due to which session was being overwritten.
Here is what was happening.
req1 and req2 both were coming at the same time. 
req1 - when request 1 comes it fetches the session data(say A) it updates session(say B).
req2 -  when request 2 comes it fetches the session data (say A) when it completes its functionality the req1 has already updated the data(say B). But still, req2 saves the data and update it to (say A).

The main question is why would everytime we store the data into the database

**Comment by HarshithaKP** [comment]
@deveshMantra Can you share the code logic which you are using ?

**Comment by deveshMantra** [comment]
@HarshithaKP 
Shared the repository link which has the code.
https://github.com/Sunbird-Ed/SunbirdEd-portal/blob/release-2.5.0/src/app/routes/accountMergeRoute.js

If you see 
https://github.com/expressjs/session/blob/b22384b712fea118f1c3eb5b0d79312ebd25e97c/index.js#L330
the shouldSave() function returns false as there is nothing to save/modified in the session. But in another condition
https://github.com/expressjs/session/blob/b22384b712fea118f1c3eb5b0d79312ebd25e97c/index.js#L340 shouldTouch() function will return !shouldSave() if store is implementing touch function(which is true in my case)
So basically for any condition, it will try to save into the session ideally this should not be the case.

**Comment by gireeshpunathil** [comment]
> req1 and req2 both were coming at the same time.

@deveshMantra - can you please elaborate this? if these are 2 separate requests (separate clients), they wont share a session ID, right? On the contrary, if they belong to same client (host IP and port), how the client is written, and what the client is expecting? I would like to know the use case better.

**Comment by deveshMantra** [comment]
@gireeshpunathil 
There are 2 separate requests from the same user.
The thing is I am using Cassandra for storing the session information. and it implements touch functionality. Now for every request, the express-session tries to save data into Cassandra.
As it has shouldsave() and storeImplementsTouch && shouldTouch(req) checks.
It assumes that data from Cassandra might be erased(TTL) and hence it tries to save the session for any case and hence overrides the session

**Comment by dougwilson** [comment]
This sounds like a bug in the store module you're using. "touch" tells the store that the session was "touched" and it should do something like keep it alive -- it should specifically *not* be writing the entire session back to the store and overriding the data.

**Comment by deveshMantra** [comment]
@dougwilson true.
Even I found same thing when I was debugging the code.
The touch method of store always sets the data it is not updating the expiry age.
When I saw other stores they are just updating expiry age and not the while session data.
